presto-hudi/src/main/java/com/facebook/presto/hudi/HudiFileSkippingManager.java

```diff
@@ -0,0 +1,312 @@
+        }
+        return candidateFileSlices;
+    }
+
+    private Map<String, List<FileSlice>> lookupCandidateFilesInMetadataTable(Map<String, List<FileSlice>> inputFileSlices, TupleDomain<ColumnHandle> tupleDomain)
+    {
+        // split regular column predicates
+        TupleDomain<HudiColumnHandle> regularTupleDomain = HudiPredicates.from(tupleDomain).getRegularColumnPredicates();
+        TupleDomain<String> regularColumnPredicates = regularTupleDomain.transform(HudiColumnHandle::getName);
+        if (regularColumnPredicates.isAll() || !regularColumnPredicates.getDomains().isPresent()) {
+            return inputFileSlices;
+        }
+        List<String> regularColumns = regularColumnPredicates.getDomains().get().entrySet().stream().map(Map.Entry::getKey).collect(Collectors.toList());
+        // get filter columns
+        List<String> encodedTargetColumnNames = regularColumns.stream().map(col -> new ColumnIndexID(col).asBase64EncodedString()).collect(Collectors.toList());
+        Map<String, List<HoodieMetadataColumnStats>> statsByFileName = metadataTable.getRecordsByKeyPrefixes(
+                encodedTargetColumnNames,
+                HoodieTableMetadataUtil.PARTITION_NAME_COLUMN_STATS, true)
+                .collectAsList()
+                .stream()
+                .filter(f -> f.getData().getColumnStatMetadata()
+                        .isPresent())
+                .map(f -> f.getData().getColumnStatMetadata().get()).collect(Collectors.groupingBy(HoodieMetadataColumnStats::getFileName));
+
+        // prune files.
+        return inputFileSlices.entrySet().parallelStream().collect(Collectors.toMap(entry -> entry.getKey(), entry -> {
+            return entry.getValue().stream().filter(fileSlice -> {
+                String fileSliceName = fileSlice.getBaseFile().map(BaseFile::getFileName).orElse("");
+                // no stats found
+                if (!statsByFileName.containsKey(fileSliceName)) {
+                    return true;
+                }
+                List<HoodieMetadataColumnStats> stats = statsByFileName.get(fileSliceName);
+                return evaluateStatisticPredicate(regularColumnPredicates, stats, regularColumns);
+            }).collect(Collectors.toList());
+        }));
+    }
+
+    private boolean evaluateStatisticPredicate(TupleDomain<String> regularColumnPredicates, List<HoodieMetadataColumnStats> stats, List<String> regularColumns)
+    {
+        if (regularColumnPredicates.isNone() || !regularColumnPredicates.getDomains().isPresent()) {
+            return true;
+        }
+        for (String regularColumn : regularColumns) {
+            Domain columnPredicate = regularColumnPredicates.getDomains().get().get(regularColumn);
+            Optional<HoodieMetadataColumnStats> currentColumnStats = stats.stream().filter(s -> s.getColumnName().equals(regularColumn)).findFirst();
+            if (!currentColumnStats.isPresent()) {
+                // no stats for column
```

Comment:
this should not happen right. can we throw here.

Reply:
i donnot think so.
The index may be expired, at this time we must return true directly instead of throwing an exception